### PR TITLE
Private Network Support

### LINKED ISSUE #931: Allowing Private networks

### PR DESCRIPTION
**Pull request by ejfitzgerald**
Implement a concept of the network node to control when/how the miner identity is checked.

-standalone now means that the network is expected to be of size 1
-private-network now means that the the network is expected to be of size > 1

Absence of both flags (default) means that it is a public network

Fixes #931

### CODE REVIEW

File: apps/constellation/main.cpp [hunk 1]
```diff
@@ -220,9 +249,10 @@ struct CommandLineArguments
...
     p.add(args.cfg.peers_update_cycle_ms, "peers-update-cycle-ms", "How fast to do peering changes",                                                uint32_t{0});
-    p.add(args.cfg.disable_signing,       "disable-signing",       "Do not sign outbound packets or verify those inbound, in trusted network",      bool{});
-    p.add(args.cfg.sign_broadcasts,       "sign-broadcasts",       "Sign and verify broadcast packets",                                             bool{});
-    p.add(args.cfg.standalone,            "standalone",            "Expect the node to run in on its own (useful for testing and development)",     false);
+    p.add(args.cfg.disable_signing,       "disable-signing",       "Do not sign outbound packets or verify those inbound, in trusted network",      false);
+    p.add(args.cfg.sign_broadcasts,       "sign-broadcasts",       "Sign and verify broadcast packets",                                             false);
+    p.add(standalone_flag,                "standalone",            "Expect the node to run in on its own (useful for testing and development)",     false);
```
Comment (wkaluza) [inline apps/constellation/main.cpp]: "to be run in on its own" - remove "in"

File: apps/constellation/main.cpp [hunk 2]
```diff
@@ -462,19 +528,13 @@ struct CommandLineArguments
     s << "external address..........: " << args.external_address << '\n';
     s << "db-prefix.................: " << args.cfg.db_prefix << '\n';
     s << "interface.................: " << args.cfg.interface_address << '\n';
-    s << "mining....................: " << ((args.cfg.block_interval_ms > 0) ? "Yes" : "No")
-      << '\n';
+    s << "mining....................: " << ToYesNo(args.cfg.block_interval_ms > 0) << '\n';
```
Comment (wkaluza) [inline apps/constellation/main.cpp:531]: :+1:

File: libs/ledger/src/protocols/main_chain_rpc_service.cpp [hunk 3]
```diff
@@ -433,5 +453,20 @@ MainChainRpcService::State MainChainRpcService::OnSynchronised(State current, St
...
+
+  // evaluate if this mining node is correct
+  bool const is_valid_miner =
+      (Mode::PUBLIC_NETWORK == mode_) ? crypto::IsFetchIdentity(block.body.miner) : true;
+  if (is_valid_miner && block.proof())
+  {
+    block_valid = true;
+  }
+
+  return block_valid;
```
Comment (wkaluza) [inline libs/ledger/src/protocols/main_chain_rpc_service.cpp:468]: How about `return is_valid_miner && block.proof();`? That's all the conditional says in this case.

File: apps/constellation/main.cpp (hunk 1 above)
Comment (wkaluza) [reply apps/constellation/main.cpp]: I'd tack on "Incompatible with -private-network" to the description.

File: apps/constellation/main.cpp [hunk 4]
```diff
@@ -220,9 +249,10 @@ struct CommandLineArguments
...
-    p.add(args.cfg.disable_signing,       "disable-signing",       "Do not sign outbound packets or verify those inbound, in trusted network",      bool{});
-    p.add(args.cfg.sign_broadcasts,       "sign-broadcasts",       "Sign and verify broadcast packets",                                             bool{});
-    p.add(args.cfg.standalone,            "standalone",            "Expect the node to run in on its own (useful for testing and development)",     false);
+    p.add(args.cfg.disable_signing,       "disable-signing",       "Do not sign outbound packets or verify those inbound, in trusted network",      false);
+    p.add(args.cfg.sign_broadcasts,       "sign-broadcasts",       "Sign and verify broadcast packets",                                             false);
+    p.add(standalone_flag,                "standalone",            "Expect the node to run in on its own (useful for testing and development)",     false);
+    p.add(private_flag,                   "private-network",       "Expect the node is to be run as part of a private network (disables bootsrap)", false);
```
Comment (wkaluza) [inline apps/constellation/main.cpp]: I'd tack on "Incompatible with -standalone" to the description.

Comment (wkaluza) [reply apps/constellation/main.cpp]: Also, "bootsrap" => "bootstrap"

**Review by wkaluza — CHANGES_REQUESTED** [review]
Looks fine, minor niggles only.

File: libs/ledger/src/protocols/main_chain_rpc_service.cpp (hunk 3 above)
Comment (ejfitzgerald) [reply libs/ledger/src/protocols/main_chain_rpc_service.cpp:468]: You are of course right. In general, I think this sort of style is easier to read actually and it makes extending the logic in the future marginally easier. I did a double check in [godbolt](https://godbolt.org/) with the following and there is no difference at `-O3` with [example.cpp.txt](https://github.com/uvue-git/fetch-ledger/files/3137109/example.cpp.txt)

File: apps/constellation/main.cpp (hunk 1 above)
Comment (ejfitzgerald) [reply apps/constellation/main.cpp]: Haha my english is terrible - whoops good spot

File: apps/constellation/main.cpp (hunk 4 above)
Comment (ejfitzgerald) [reply apps/constellation/main.cpp]: ✅

File: libs/ledger/src/protocols/main_chain_rpc_service.cpp (hunk 3 above)
Comment (wkaluza) [reply libs/ledger/src/protocols/main_chain_rpc_service.cpp:468]: That's all right then.

**Review by wkaluza — APPROVED** [review]

**Pull request merged**